File: leetcode/0029/divide.swift
import Foundation

func divide(_ dividend: Int, _ divisor: Int) -> Int {
    var sign = 1
    if divisor < 0 && dividend < 0 {
        sign = 1
    } else if divisor < 0 || dividend < 0 {
        sign = -1
    }
    
    let dividend = abs(dividend)
    let divisor = abs(divisor)
    var remain = dividend
    var result = 0
    while remain >= divisor {
        var cur = 1
        var div = divisor
        while div + div < remain {
            cur += cur
            div += div
        }
        remain -= div
        result += cur
    }
    
    result = sign == -1 ? 0 - result : result
    
    let max: Double = pow(2, 31)
    return result >= Int(max) ? Int(max) - 1 : result
}

File: leetcode/0029/divide.py
def divide(dividend: int, divisor: int) -> int:
    sign = 1
    if divisor < 0 and dividend < 0:
        sign = 1
    elif divisor < 0 or dividend < 0:
        sign = -1

    divisor = abs(divisor)
    dividend = abs(dividend)

    remain = dividend  # 余数
    result = 0  # 商
    while remain >= divisor: 
        cur = 1 # 倍增商
        div = divisor # 倍增值
        while div + div < remain:
            cur += cur 
            div += div 
        remain -= div  # 余数递减
        result += cur  # 商值累计
    
    if sign==-1:
        result = -result

    return 2**31-1 if result >= 2**31 else result

print(divide(10, 3))